### ISSUE TITLE
Q4 2020 roadmap

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno#7915]
Update 2020-12-30: we're not gonna upgrade to Tokio 0.3, but instead go straight to Tokio 1.0 (#8779), however we're still blocked by a few dependencies. Given that ResourceTable refactor eliminated most of the quirks/hangs and year is nearing the end I'm gonna close this issue as resolved.

Update 2020-12-17: rewrite of ResourceTable landed in #8512 and migration to Tokio 0.3 is in progress in #8779. That should
allow to close #8416 and most of issues listed in `bugs`.

Update 2020-11-23: we're on track with most of items listed in this issue and I hope we can wrap them up before end of the year. Unfortunately we won't be able to address some points from "features" and "maybe" so I'm dropping them from the list. 

Here's the roadmap for `cli` and `core` for the last quarter of this year.

Refactors:
- [x] finish compiler/bundler refactor (#7225, #4752)
- [x] finish TLA implementation https://github.com/denoland/deno/pull/7946
- [x] refactor module loader in core (#3736)
- [ ] rewrite ResourceTable (#8416)

_maybe_
- [x] refactor file fetcher infrastructure - resolve cache busting problem (#6116, #5631, #5253, others)
- [ ] finish op crates refactor - tests, remaining web APIs in crates
- [ ] <s>rewrite "prepareStackTrace" API (#7644)</s>
- [ ] <s>better error messages (using .context())</s>

Bugs:
- [ ] harden subprocess implementation (#7087, #3014, #5621, #7713, #7727)
- [ ] harden workers implementation (#7620, #6320, #5334, #4728)
- [ ] harden file watcher implementation (#7590, #7627)
- [ ] harden WS implementation (#7457)
- [x] harden HTTPS server implementation (#8107)

Features:
- [x] MVP for "deno lsp" - provide minimal version of LSP that would replace logic in vscode_deno (#8400)
- [ ] <s>structured clone for workers (#6887)</s>
- [ ] <s>unref timers (#6141)</s>
- [ ] <s>doctest support (#7916)</s>
- [x] deno compile (https://github.com/denoland/deno/pull/8539)
- [x] provide library crate (#7928, #2633)
_maybe_
- [ ] <s>WASM imports (#5609)</s>
- [ ] <s>JSON imports (#7623)</s>
- [ ] <s>extensible testing infrastructure (#7840)</s>

_All of these are being worked on, and they tend to be at the complex end of the spectrum. If there are small bits where contribution can be easy, we will try to break it out as a separate issue._

__Keep in mind this is a roadmap for the core team and contributors' proposals are currently not included in it.__ I'm more than happy to coordinate with contributors to include more propositions in the roadmap that would be tackled by the community.

**Comment by chenxyzl** [comment]
maybe gpu support for ML?

**Comment by iykekings** [comment]
@bartlomieju I have moved #6124 to #7916

**Comment by newtack** [comment]
Is TranspileOnly use with swc still slated for 1.5?

**Comment by kitsonk** [comment]
@newtack that would be part of #4752

**Comment by ebebbington** [comment]
great news that https://github.com/denoland/deno/issues/7727 is on the roadmap (as part of the bugs section), got a browser testing module i want to release but unsure if it'll end up being possible to fix the problem i have

**Comment by caspervonb** [comment]
My big items:

- Revamp the REPL, turn it into a general purpose debugger.
- New parallel capable test runner in Rust with multiple reporters (pretty, json, custom).
- Best in class test coverage cli with multiple reporters (pretty, json, custom).
